search.go

```diff
@@ -1,175 +1,152 @@
+// WithSize sets the request size, limiting the number of documents returned.
 // A negative value for size will be ignored and not included in the SearchRequest.Source.
-func (r *SearchRequest) SetSize(n int) *SearchRequest {
+func (r *SearchRequest) WithSize(n int) *SearchRequest {
 	r.Size = n
 	return r
 }
 
-// AddSort to the current list of [search.Sort]s on the request.
-func (r *SearchRequest) AddSort(sort ...*search.Sort) *SearchRequest {
+// WithSorts to the current list of [Sort]s on the request.
+func (r *SearchRequest) WithSorts(sort ...*Sort) *SearchRequest {
 	r.Sort = append(r.Sort, sort...)
 	return r
 }
 
-// SetQuery to be performed by the SearchRequest.
-func (r *SearchRequest) SetQuery(q search.Query) *SearchRequest {
+// WithQuery to be performed by the SearchRequest.
+func (r *SearchRequest) WithQuery(q Query) *SearchRequest {
 	r.Query = q
 	return r
 }
 
-// Do executes the SearchRequest using the provided opensearch.Client.
-// If the request is executed successfully, then a SearchResponse will be returned.
-// An error can be returned if
+// SearchResponse is a domain model union response type across all supported OpenSearch versions.
+// Currently supported versions are:
 //
-//   - The SearchRequest source cannot be created
-//   - The source fails to be marshaled to JSON
-//   - The OpenSearch request fails to executed
-//   - The OpenSearch response cannot be parsed
-func (r *SearchRequest) Do(ctx context.Context, client *opensearch.Client) (*SearchResponse, error) {
-	bodyBytes, jErr := json.Marshal(r)
-	if jErr != nil {
-		return nil, jErr
-	}
-
-	osResp, rErr := opensearchapi.SearchRequest{
-		Index: r.Index,
-		Body:  bytes.NewReader(bodyBytes),
-	}.Do(ctx, client)
-
-	if rErr != nil {
-		return nil, rErr
-	}
-
-	var respBuf bytes.Buffer
-	if _, err := respBuf.ReadFrom(osResp.Body); err != nil {
-		return nil, err
-	}
-
-	resp := &SearchResponse{
-		StatusCode: osResp.StatusCode,
-		Header:     osResp.Header,
-	}
-
-	if err := json.Unmarshal(respBuf.Bytes(), &resp); err != nil {
-		return nil, err
-	}
-
-	return resp, nil
-}
-
-// SearchResponse wraps the functionality of [opensearchapi.Response] by supporting request parsing.
+//	-OpenSearch2
 type SearchResponse struct {
+	// StatusCode of the http request
 	StatusCode int
-	Header     http.Header
-	Took       int       `json:"took"`
-	TimedOut   bool      `json:"timed_out"`
-	Shards     ShardMeta `json:"_shards,omitempty"`
-	Hits       Hits      `json:"hits"`
-	Error      *Error    `json:"error,omitempty"`
+
+	// Header details returned by OpenSearch
+	Header http.Header
+
+	// Took the time in Milliseconds OpenSearch took to execute the query
+	Took int
+
+	// TimedOut true if the request timed out
+	TimedOut bool
+
+	// Shards [ShardMeta] counts of the shards used in the request processing
+	Shards ShardMeta
+
+	// Hits are the results of the [Query]
+	Hits Hits
+
+	// Error if OpenSearch failed but responded with errors
+	Error *Error
```

Comment:
Same question here about pointer vs value.  `Error` is a plain-old-data struct.

Reply:
Conversed offline. The intent of the pointer is to treat it like an `error` allowing `if resp.Error != nil { /* process */}` checks.